site/app/controllers/ManageSessionsController.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace app\controllers;
+
+use app\libraries\Core;
+use app\libraries\response\WebResponse;
+use app\libraries\response\RedirectResponse;
+use Symfony\Component\Routing\Annotation\Route;
+use app\entities\Session;
+use app\repositories\SessionRepository;
+
+/**
+ * Class ManageSessionsController
+ * @package app\controllers
+ */
+class ManageSessionsController extends AbstractController {
+    /**
+     * ManageSessionsController constructor.
+     *
+     * @param Core $core
+     */
+    public function __construct(Core $core) {
+        parent::__construct($core);
+    }
+    /**
+     * Show manage sessions page
+     *
+     * @Route("/manage_sessions", methods={"GET"})
+     * @return WebResponse
+     */
+    public function showSessionsPage(): WebResponse {
+        /** @var SessionRepository $repo */
+        $repo = $this->core->getSubmittyEntityManager()->getRepository(Session::class);
+        $user_sessions = $repo->getAllByUser($this->core->getUser()->getId());
+        return new WebResponse(
+            'ManageSessions',
+            'showSessionsPage',
+            $user_sessions
+        );
+    }
+
+    /**
+     * Terminate a session
+     *
+     * @Route("/manage_sessions/logout", methods={"POST"})
+     * @return RedirectResponse
+     */
+    public function logoutFromSession(): RedirectResponse {
+        if (isset($_POST["session_id"])) {
+            $session_id = $_POST["session_id"];
+        }
+        else {
+            $this->core->addErrorMessage("Session id not provided.");
+            return new RedirectResponse($this->core->buildUrl(["manage_sessions"]));
+        }
+        $session = $this->core->getSubmittyEntityManager()->getRepository(Session::class)
+            ->findOneBy(['session_id' => $session_id]);
+        if (empty($session)) {
+            $this->core->addErrorMessage("Session doesn't exist.");
+            return new RedirectResponse($this->core->buildUrl(["manage_sessions"]));
+        }
+        if ($session->getUserId() === $this->core->getUser()->getId()) {
+            if ($this->core->getCurrentSessionId() === $session_id) {
+                return new RedirectResponse($this->core->buildUrl(["authentication", "logout"]));
+            }
+            $this->core->getSubmittyEntityManager()->remove($session);
+            $this->core->getSubmittyEntityManager()->flush();
+            $this->core->addSuccessMessage("Session terminated successfully.");
+        }
+        else {
+            $this->core->addErrorMessage("You don't have permission to terminate that session.");
+        }
```

Comment:
Can we make the error the same whether the session doesn't exist or it doesn't belong to user? This could let a user start guessing other session IDs. There isn't anything harmful that could be done with a session ID without the JWT secret but I think it's safer to just give the same error message or edit the logic so there is only one return with an error.

Reply:
It has been changed now to say 'Session doesn't exist' for both cases.